### PR TITLE
[watermarking] golden tests for backwards compatibility

### PR DESCRIPTION
**Pull request by TQCB**
## Purpose
Watermarking is a paired contract between a generation process and a detector, so we want to make sure this contract is maintained. Watermarking is also used for regulatory compliance, and we have to detect the watermarking signal on all text that we have watermarked, even if it was with an outdated algorithm.

This PR adds a suite of golden tests for watermarking that forces this contract in testing. 

As mentioned in documentation in the PR, these tests do not instantiate the model runner or `GPUWatermarkSampler`, and therefore do not cover:
- model logits
- top-k/top-p
- temperature processing
- request/batch state
- tokenizer <-> text round trips
- CUDA kernels
- distributed execution
- Speculative-decoding draft/accept/recovery/bonus

We consider that these boundaries need sampler/kernel parity tests and shouldn't be handled by these goldens

### CODE REVIEW

**Comment by simon-veitner-redhat** [comment]
LGTM

File: tests/watermarking/README.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
+# Watermarking goldens
+
+These goldens freeze the CPU reference path: the generator and detector for each
+configured watermarking algorithm and PRF combination, over a deterministic
+logit fixture. The current matrix uses the Philox PRF. Regeneration refuses an
+incomplete algorithm/PRF matrix.
+They are a backwards-compatibility check, not a check of GPU numerics. A
```
Comment (NickLucche) [inline tests/watermarking/README.md:7]: do we need readme in tests?

File: tests/watermarking/golden_candidates.py [hunk 2]
```diff
@@ -0,0 +1,1224 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+"""Deterministic watermarking golden candidates; see README.md."""
+
+import json
+import math
+import platform
+from collections.abc import Mapping
```
Comment (NickLucche) [inline tests/watermarking/golden_candidates.py:9]: there's a lot of structure on this file for a test, which makes it huge; do we really need to define all these dataclasses?

**Review by LucasWilkinson — COMMENTED** [review]
Ditto what @NickLucche said; this is alot of code, can we trim this down at all ?

**Comment by simon-veitner-redhat** [comment]
i'll try to trim it down tomorrow!
